### DISCUSSION TITLE
programmatically get homebrew formula-file

### DISCUSSION THREAD

**Discussion by umlaeute** [Homebrew/discussions#4569] [Tap maintenance and Homebrew development] [answered]
### Output of `brew config`

```shell
--
```

### Output of `brew doctor`

```shell
--
```

### Description of issue

i originally asked this [on stackoverflow](https://stackoverflow.com/questions/76345952/), but was told to cross-post here. so...

i would like to do some programmatic patching of homebrew's formula files for named packages (e.g. `aalib`). for this i obviously need to have have the formula file first (some `aalib.rb` file).

for standard formulae this is pretty straight-forward (just get them from https://github.com/Homebrew/homebrew-core/).
Things start to get trickier, once you consider additional (user-defined) taps.

i can think of a number of (what i consider) hacks, to retrieve the formula:
- just get the formula directly from `https://github.com/Homebrew/homebrew-core/blob/master/Formula/${name}.rb` 
   - this obviously ignores any user-defined taps :-/
- iterate over all locally activated taps, and check whether their local checkout has any matching formula
  ```sh
  for tap in $(brew tap); do
    f=$(brew --repository "${tap}")/Formula/"${name}.rb"
    test -e "${f}" && echo "${f}"
  done
  ```
  - this might give us multiple formulae-file (and i cannot just take the first formula file, as `brew tap` will return taps in alphabetical order, but `brew install aalib` will prefere the `homebrew/core` tap), :-/

- retrieve the *content* of a given formula and write it to a file
  ```
  EDITOR=cat brew edit "${name}" > "${name}.rb"
  ```
  - this has one useless indirection (which could be mitigated by just using the `stdout` for patching the formula definition)
  - `brew edit` also outputs a line like "Editing /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/aalib.rb" to the stdout before actually opening the formula file. of course this is easily removed, but it makes me think that this interface is brittle and likely to break with future homebrew updates :-/

so: is there a canonical way to get the actual formula-file (or its content), that would be used to install a given formula (using only its name)?

PS: my patched formula files are not going to overwrite the original ones, so I would actually be fine with just the contents of the file (if there is a stable cmdline API to get it). otoh it would be a bonus to also know which tap a given formula originated from.

**Comment by carlocab** [comment, marked as answer]
Try
```sh
brew ruby -e "puts '$name'.f.path"
```
For example:
```
❯ name=hello; brew ruby -e "puts '$name'.f.path"
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/hello.rb
```
